### PR TITLE
Add batch parameters and export data in batches

### PR DESCRIPTION
**Pull request by surajreddy**
### What it does and why:
- Retrieves data in batches to reduce load on the server, and to incrementally export data.

### Checklist:

- [x] Testing & QA: Passes tests and linting.
- [X] :tophat:: I have manually tested these changes.

### CODE REVIEW

File: exe/shopify_transporter [hunk 1]
```diff
@@ -35,10 +35,17 @@ module ShopifyTransporter
     method_option :object, type: :string, required: true,
       desc: 'The object type (customer, product, or order) to export.',
       enum: %w(customer product order)
+    method_option :first_id, type: :numeric, required: true,
+      desc: 'The starting id of the object type you are trying to export.'
```
Comment (JordanVerasamy) [inline exe/shopify_transporter]: These descriptions are not very descriptive. I'm not sure how to improve them, but I definitely think they need attention.

Comment (fahia) [reply exe/shopify_transporter]: I'm not sure what's the best wording here. Might be worth getting into the range query args in the new wiki.

Comment (surajreddy) [reply exe/shopify_transporter]: Based on some input from Jenna, I'll change this to:

```
first_id

The first product_id, customer_id, or order number for the range of objects you are trying to export.


last_id

The last product_id, customer_id, or order number for the range of objects you are trying to export.


batch_size

The number of records that are retrieved from the source platform at one time. Use a smaller batch size if you are running into performance issues extracting your data.
```

Comment (JordanVerasamy) [reply exe/shopify_transporter]: Great updates.

**Comment by JordanVerasamy** [comment]
The changes are good. Once this is tophatted and stuff I'll feel comfortable merging this.

File: lib/shopify_transporter/exporters/magento/customer_exporter.rb [hunk 2]
```diff
@@ -26,8 +26,13 @@ def with_attributes(base_customer)
         end
 
         def base_customers
-          result = @client.call(:customer_customer_list, filters: nil).body
-          result[:customer_customer_list_response][:store_view][:item] || []
+          Enumerator.new do |enumerator|
+            @client.call_in_batches(method: :customer_customer_list, batch_index_column: 'customer_id').each do |batch|
+              result = batch.body[:customer_customer_list_response][:store_view][:item] || []
+              result = [result] unless result.class == Array
```
Comment (mimperatore) [inline lib/shopify_transporter/exporters/magento/customer_exporter.rb]: More idiomatic: `result is_a? Array`, or:

```
result = [*batch.body[:customer_customer_list_response][:store_view][:item]]
```

File: exe/shopify_transporter [hunk 3]
```diff
@@ -35,10 +35,18 @@ module ShopifyTransporter
...
+    method_option :last_id, type: :numeric, required: true,
+      desc: 'The last product_id, customer_id, or order_id for the range of objects you are trying to export.'
+    method_option :batch_size, type: :numeric, default: 1000,
+      desc: 'The number of records that are retrieved from the source platform at one time. '\
+        'Use a smaller batch size if you are running into performance issues extracting your data.'
 
     desc 'export', 'Exports objects from a third-party platform into a format compatible for conversion.'
     def export
-      exporter = Exporters::Exporter.new(options[:config], options[:object])
+      batch_config = options.select { |(key, _)| %w(first_id last_id batch_size).include? key }
```
[the remark this answers is not in the document]
Comment (surajreddy) [reply exe/shopify_transporter]: We don't have access to slice with vanilla Ruby, we will need to bring in `require 'active_support/core_ext/string'`

I think that's a good idea/I'll go ahead and do this.

File: lib/shopify_transporter/exporters/magento/soap.rb [hunk 4]
```diff
@@ -7,21 +7,55 @@ module ShopifyTransporter
...
+        MAX_RETRIES = 4
+
+        def initialize(hostname: '', username: '', api_key: '', batch_config:)
           @hostname = hostname
           @username = username
           @api_key = api_key
+          @batch_config = batch_config
+        end
+
+        def call(method, params)
```
Comment (mimperatore) [inline lib/shopify_transporter/exporters/magento/soap.rb]: To be consistent with the interface of `call_in_batches`, should we make this also `method:, params:`?

Comment (mimperatore) [reply lib/shopify_transporter/exporters/magento/soap.rb]: Or conversely changing `call_in_batches` _not_ to use named parameters?  Most methods are not using names params, so perhaps that should be the convention we go with.

File: spec/shopify_transporter/exporters/magento/soap_spec.rb [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+
+            attempt = 0
+            expect(mock_client).to receive(:call).with(
+              :test_call,
+              message: {
+                session_id: '123',
+              },
+            ) do |args|
+              attempt += 1
+              if attempt < 5
```
Comment (mimperatore) [inline spec/shopify_transporter/exporters/magento/soap_spec.rb]: Missed this one, I think.

File: lib/shopify_transporter/exporters/magento/soap.rb (hunk 4 above)
Comment (surajreddy) [reply lib/shopify_transporter/exporters/magento/soap.rb]: Done!

**Pull request merged**